accepted/2021/package-scoring/PackageScoring.md

```diff
@@ -0,0 +1,259 @@
+  - Demonstrate to people that packages are impactful, valuable, and have traction in the ecosystem.
+  - Help package authors sustain their work by incentivizing their time with raising donations/funding.
+
+There are also many **jobs-to-be done** when browsing for packages to use in projects or deciding to keep in your software supply chain. Here is a small list of jobs this concept will help developers with.
+
+1. Determine whether a package is recognized by the ecosystem based on its **popularity and community** score.
+2. Understand if a package is actively maintained or abandonded based on its **maintenance, quality, and community** score.
+3. Prevent installing a package that contains a known vulnerability that has not yet been resolved based on it's **maintenance and quality** score.
+4. See how they can get involved with the development of the package based on its **community** score.
+5. Assess whether the package & its dependencies are compatible with the latest .NET Frameworks & tooling based on it's **quality** score.
+
+The creation of a score for each of these categories and an overall score per package will help assess the health of the .NET Ecosystem & NuGet.
+
+A developer for example can assess the score of a package to make a better decision as to if the package is worth including in their project. A team at a company might even take this score into their dependency policy to say they only accept a package score of 80 or above because they have stricter requirements.
+
+## Concepts
+
+**NuGet Gallery**
+
+![](PackageScoring.png)
+
+![](PackageScoringList.png)
+
+**Visual Studio**
+
+![](PackageScoringVisualStudio.png)
+
+**.NET CLI / Tool**
+
+![](PackageScoringTool.png)
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+When a .NET developer creates a NuGet package and pushes it to NuGet.org, it will be scored on the following metrics totaling 25 points per category up to an overall score of 100.
+
+**Popularity**:
+
+- Total Weekly Downloads
+- Number of Dependents
+- Total GitHub Stars
+- Total GitHub Forks
+- Total GitHub Contributors
+
+Based on these factors, this score can fall into a bucket that helps the user understand its popularity on NuGet such as “limited” or “key ecosystem project”.
+
+**Quality**:
+
+- Follow NuGet conventions
+- Provides documentation
+- Supports multiple platforms (if it can)
+- Pass static analysis (reproducible builds, api compatibility, new feature adoption)
+- Support up-to-date dependencies at creation time
+
+Based on these factors, this score can fall into a bucket that helps the user understand its quality on NuGet such as “low” or “high”.
+
+**Maintenance**:
+
+- Open GitHub Issues
```

Comment:
This would be hard to measure. Huge active projects have thousands of open Issues. Stale projects would also have many open issues.

Reply:
How some other ecosystems have accomplished this is by having a reasonable threshold or optimal stop to achieve the maximum score for these ratios. Definitely shouldn't be punished for having issues, we all have em! 😄

Reply:
What is useful to track is the open/closed ratio, for established projects this usually gives you a good idea of how serious their maintenance effort is.

Reply:
Here's how npm scores issues: https://github.com/npms-io/npms-analyzer/blob/master/lib/scoring/score.js#L70-L84

Reply:
@JonDouglas commit frequency is a good one.

Reply:
Although if you take a project like Reactive Extensions - it's very mature, and the commit frequency is quite low https://github.com/dotnet/reactive/commits/main

Reply:
> Definitely shouldn't be punished for having issues, we all have em! 😄

Speak for yourself 😜

Reply:
It might be useful to define a set of archetypal packages with various characteristics, and consider how they would be scored under these metrics. For example, what do we want to score to look like for a small, tightly scoped, well maintained, mature project? Or an abandoned package built out of the same repo as more mature and actively developed packages?

Reply:
@mhutch Absolutely. We'll be able to run feasibility dry runs & provide tooling for anyone to test locally for what can be scored up front. These types of examples should put the proposal to the test & there's many more categories of metadata we can include and run through.